#67 Add an XML Stylesheet

kirstyburgoine opened the pull request:
### Issue Number
https://github.com/GoogleChromeLabs/wp-sitemaps/issues/67 - Add an XML Stylesheet (XSL file)
This also addresses https://github.com/GoogleChromeLabs/wp-sitemaps/issues/77  - Make CSS for XSL file filterable

### Description
Adds an .xsl stylesheets file to style all of the sitemaps 

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Enhancement (change which improves an existing feature. E.g., performance improvement, docs update, etc.)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
View any of the stylesheets listed below to see them laid out in a table with some basic styles applied
- http://your-local/sitemap.xml
- http://your-local/sitemap-posts-post.xml
- http://your-local/sitemap-taxonomies-category.xml
- http://your-local/sitemap-users.xml
(You may need to flush your redirects locally)

### Screenshots

**Sitemap Index**

<img width="1132" alt="Screenshot 2019-11-21 at 17 35 14" src="https://user-images.githubusercontent.com/182015/69362242-0fe46a00-0c86-11ea-907d-1219bb4fbdba.png">

**Pages Sitemap**

<img width="1132" alt="Screenshot 2019-11-21 at 17 35 00" src="https://user-images.githubusercontent.com/182015/69362243-107d0080-0c86-11ea-9c99-718fd0b93604.png">

### Acceptance criteria
- [x] My code follows WordPress coding standards.
- [x] I have performed a self-review of my own code.
- [x] If the changes are visual, I have cross browser / device tested.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added test instructions that prove my fix is effective or that my feature works.

kirstyburgoine committed:
67: Base implementation of xsl file

kirstyburgoine committed:
67: Add sitemap index filter, remove hardcoded url

kirstyburgoine committed:
67: Add some basic styles

kirstyburgoine committed:
67: Remove unused url

kirstyburgoine committed:
67: Add sitemap styles to index

kirstyburgoine committed:
67: Change sitemap.xsl to sitemap.xsl.php

kirstyburgoine committed:
67: wip rewrite rule for sitemap.xsl.php

kirstyburgoine committed:
67: Add comment

kirstyburgoine committed:
Merge branch 'master' into feature/67-xml-stylesheet

kirstyburgoine committed:
67: Update comment with @todo

joemcgill commented on inc/class-core-sitemaps-renderer.php:
```diff
@@ -48,6 +72,13 @@ public function render_index( $sitemaps ) {
 		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 		echo $sitemap_index->asXML();
+
+		/**
```
I don't think this is the right location for this filter. Did you mean for it to be in `get_sitemap_stylesheet_url()` instead?

joemcgill commented on inc/class-core-sitemaps-renderer.php:
```diff
@@ -75,5 +106,12 @@ public function render_sitemap( $url_list ) {
 		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 		echo $urlset->asXML();
+
+		/**
```
Same as above. You want to put the filter inside the function that returns the stylesheet location rather than here. This function already echoes the created XML so it doesn't really return anything that would be useful to filter.

joemcgill commented on inc/class-core-sitemaps.php:
```diff
@@ -96,6 +97,13 @@ public function setup_sitemaps() {
+	/**
+	 * Provide rewrite for the xsl stylesheet.
+	 */
+	public function xsl_stylesheet_rewrite() {
```
Ah, I see what's happening here.

The rewrite rule will need to direct to something like `index.php?sitemap-xsl=1`, and then you can add a method to the `Core_Sitemaps` class that works similar to how `Core_Sitemaps_Index::render_sitemap()` works, where you hook to the `template_redirect` action with a function that looks for the existence of the `sitemap-xsl` param—which you'll need to add an `add_rewrite_tag` for—and then include your sitemap.xsl.php file programmatically and `exit`.

kirstyburgoine committed:
67: Move filters

kirstyburgoine committed:
67: add xsl as class and fix rewrites

This still isn’t working quite correctly. Rewrites work properly but the xsl is rendering directly to the page instead of being used to style sitemap.xml

kirstyburgoine committed:
Merge branch 'master' into feature/67-xml-stylesheet

kirstyburgoine committed:
67: Delete unused file and update doc

kirstyburgoine committed:
67: Add escaping, i18n & CSS filter to XSL content

kirstyburgoine committed:
67: Change xsl url and update filter name

kirstyburgoine committed:
67: Add check for stylesheet query_var

kirstyburgoine committed:
67: Move query_var to stylesheet renderer

kirstyburgoine committed:
67: Remove unneeded / from xsl url

kirstyburgoine committed:
Merge branch 'master' into feature/67-xml-stylesheet

kirstyburgoine committed:
67: Update doc

kirstyburgoine committed:
67: Remove echo from render_stylesheet()

kirstyburgoine committed:
67: Update printf to sprintf

joemcgill commented on inc/class-core-sitemaps-stylesheet.php:
```diff
@@ -0,0 +1,156 @@
+			 *
+			 * @param string $xsl Full content for the xml stylesheet.
+			 */
+			return apply_filters( 'core_sitemaps_stylesheet_content', $xsl );
```
Instead of returning here, you'll need to echo the XSL and then `exit` so WordPress doesn't continue with it's templating logic.

kirstyburgoine committed:
67: Move escsped content out of xml output

Commit:
Echo out the XSL file and die.

kirstyburgoine committed:
67: Update escapde content and small tidy

kirstyburgoine committed:
67: Add index styles

kirstyburgoine committed:
67: Tidy up of docs

kirstyburgoine committed:
67: Combine xsl to output content based query_var

kirstyburgoine committed:
67: Tidy

kirstyburgoine committed:
Revert "67: Combine xsl to output content based query_var"

This reverts commit ae5ad7759f8d82974de0723dd8f6dd8359c04cd4.

kirstyburgoine committed:
67: Tidy

kirstyburgoine committed:
67: phpcs

kirstyburgoine committed:
Merge branch 'master' into feature/67-xml-stylesheet

kirstyburgoine commented on inc/class-core-sitemaps-stylesheet.php:
```diff
@@ -0,0 +1,234 @@
+	/**
+	 * Returns the escaped xsl for the index sitemaps.
+	 */
+	public function stylesheet_index_xsl() {
```
In commit https://github.com/GoogleChromeLabs/wp-sitemaps/pull/75/commits/ae5ad7759f8d82974de0723dd8f6dd8359c04cd4 I combined both of these so that there was only one output and one filter, and the content within changed based on the query_var. I later realised though that if anyone was going to filter this content they would also need to add in checks for whether it was the index sitemap or not so decided it was better to keep them completely seperate

kirstyburgoine committed:
67: Add domain args back

joemcgill commented on inc/class-core-sitemaps-stylesheet.php:
```diff
@@ -0,0 +1,234 @@
+			header( 'Content-type: application/xml; charset=UTF-8' );
+
+			if ( 'xsl' === $stylesheet_query ) {
+				// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped -- All content escaped below.
```
👍

joemcgill commented on inc/class-core-sitemaps.php:
```diff
@@ -96,6 +98,27 @@ public function setup_sitemaps() {
+	public function index_xsl_stylesheet_rewrite() {
+		add_rewrite_rule( '^sitemap-index\.xsl$', 'index.php?stylesheet=index', 'top' );
+
+		$stylesheet = new Core_Sitemaps_Stylesheet();
```
These two lines are redundant with the ones above in `xsl_stylesheet_rewrite()` above since both are going to fire on the `init` action. I would combine these two methods int a single method that adds rewrite rules for both stylesheet types.

joemcgill reviewed (changes requested):
This looks good and is working great now, @kirstyburgoine. Nice work! There's one change I'd like to see with the way rewrites are handled (comment below) and then I think this is good to merge.

kirstyburgoine committed:
67: Combine rewrite functions into one
